packages/zoe/contracts/helpers/bondingCurves.js

```diff
@@ -0,0 +1,77 @@
+import { natSafeMath } from './safeMath';
+
+const { add, subtract, multiply, divide } = natSafeMath;
+
+/**
+ * Contains the logic for calculating how many units should be given
+ * back to the user in exchange for what they sent in. It also
+ * calculates the new units of the assets in the pool. Reused in
+ * several different places, including to check whether an offer is
+ * valid, getting the current price for an asset on user request, and
+ * to do the actual reallocation after an offer has been made.
+ * @param  {object} zoe - the contract facet of Zoe
+ * @param  {assay[]} assays - an array of assays
+ *
+ * Once the function is made, it has the following parameters:
+ * @param  {units[]} poolUnitsArray - an array of the current units in the
+ * liquidity pool
+ * @param  {units} unitsIn - the units sent in by a user
+ * @param  {number} feeInTenthOfPercent=3 - the fee taken in tenths of
+ * a percent. The default is 0.3%. The fee is taken from unitsIn
+ */
+
+export const makeCalculateConstProductFn = (zoe, assays) => {
+  const unitOpsArray = zoe.getUnitOpsForAssays(assays);
+
+  return (poolUnitsArray, unitsIn, feeInTenthOfPercent = 3) => {
+    const assayIn = unitsIn.label.assay;
+    if (assayIn !== assays[0] && assayIn !== assays[1]) {
+      throw new Error(`unitsIn ${unitsIn} were malformed`);
+    }
+    const IN_INDEX = assayIn === assays[0] ? 0 : 1;
+    const OUT_INDEX = 1 - IN_INDEX;
+
+    // Constant product invariant means:
+    // tokenInPoolE * tokenOutPoolE =
```

Comment:
Comment terminology should match code terminology.

This is a fine first statement of the invariant, as a first approximation involving real numbers. But this invariant, with exact equality, is impossible with natural numbers. However, it is not good enough to simply say "as close as possible" to this. You should state the actual invariant your code is supposed to uphold, and which you believe it does uphold. I think you'll need a separate variable with the target product, rather than assuming that the actual product of the two balances is the target. The actual product should be as close as possible to the target product while remaining >= to the target product.

Reply:
I rewrote this part to better match the uniswap code, which handles rounding